### PR TITLE
Remove cluster update task when task times out

### PR DESCRIPTION
**Pull request by ywelsch**
Fixes an issue where the cluster service does not remove an update task from its internal data structures that are used for batching cluster state updates.

Relates #21568

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java [hunk 1]
```diff
@@ -466,12 +468,29 @@ public void submitStateUpdateTask(final String source, final ClusterStateUpdateT
...
                     for (UpdateTask<T> task : updateTasks) {
                         if (task.processed.getAndSet(true) == false) {
-                            logger.debug("cluster state update task [{}] timed out after [{}]", source, config.timeout());
-                            task.listener.onFailure(source, new ProcessClusterEventTimeoutException(config.timeout(), source));
+                            logger.debug("cluster state update task [{}] timed out after [{}]", source, timeout);
+                            toRemove.add(task);
+                        }
+                    }
+                    if (toRemove.isEmpty() == false) {
+                        ClusterStateTaskExecutor<T> clusterStateTaskExecutor = toRemove.iterator().next().executor;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java]: Nit: `toRemove.iterator().next().executor` -> `toRemove.get(0).executor`

**Review by jasontedor — APPROVED** [review]
LGTM. I left a minor comment.

**Comment by jasontedor** [comment]
Thanks @ywelsch. I think is good to address the problem, and we can think longer term about how to simplify this specific code in the cluster service.

**Comment by ywelsch** [comment]
Thanks @jasontedor. I'll leave it running on CI tonight and backport tomorrow.

**Comment by ywelsch** [comment]
backport complete (5.x: daf7dbe, 5.0: bbf52d1, 2.4: 161c65a)

**Pull request merged**